node/src/database/rocksdb.rs

```diff
@@ -7,130 +7,350 @@
+
+        Self {
+            rocksdb: Arc::new(
+                rocksdb_lib::OptimisticTransactionDB::open_cf(
+                    &opts,
+                    Path::new(&path),
+                    [CF_LEDGER, CF_CANDIDATES],
+                )
+                .expect("should be a valid database"),
+            ),
+        }
+    }
 
-    fn view<F>(&mut self, f: F) -> Result<()>
+    fn view<F>(&self, f: F) -> Result<()>
     where
-        F: FnOnce(&Transaction) -> Result<()>,
+        F: for<'a> FnOnce(Self::T<'a>) -> Result<()>,
     {
-        // Create read-only transaction
-        let mut tx = self.begin_tx(TxType::ReadOnly);
+        // Create a new read-only transaction
+        let tx = self.begin_txn(TxType::ReadOnly);
 
-        // If f returns err, no commit will be applied into backend
-        // storage
-        f(&tx)?;
-
-        // Release tx resources
-        tx.close();
+        // Execute all read-only transactions in isolation
+        f(tx)?;
 
         Ok(())
     }
 
-    fn update<F>(&mut self, f: F) -> Result<()>
+    fn update<F>(&self, f: F) -> Result<()>
     where
-        F: FnOnce(&Transaction) -> Result<()>,
+        F: for<'a> FnOnce(&Self::T<'a>) -> Result<()>,
     {
         // Create read-write transaction
-        let mut tx = self.begin_tx(TxType::ReadWrite);
+        let tx = self.begin_txn(TxType::ReadWrite);
 
         // If f returns err, no commit will be applied into backend
         // storage
         f(&tx)?;
 
-        // Apply changes
+        // Apply changes in atomic way
         tx.commit()?;
 
-        // Release tx resources
-        tx.close();
-
         Ok(())
     }
 
     fn close(&mut self) {}
 }
 
-pub struct Transaction {}
+pub struct Transaction<'db, DB: DBAccess> {
+    inner: rocksdb_lib::Transaction<'db, DB>,
+    access_type: TxType,
 
-impl Tx for Transaction {
-    // Read-only transactions.
-    // fn fetch_block_header(&self, hash: &[u8]) -> Result<&Header> {
-    //     // Code for fetching block header here
-    //     // ...
-    // }
+    candidates_cf: &'db ColumnFamily,
+    ledger_cf: &'db ColumnFamily,
+    snapshot: SnapshotWithThreadMode<'db, DB>,
+}
 
-    // fn fetch_block_txs(&self, hash: &[u8]) -> Result<Vec<ContractCall>> {
-    //     // Code for fetching block transactions here
-    //     // ...
-    // }
+impl<'db, DB: DBAccess> Tx for Transaction<'db, DB> {
+    fn store_block(&self, b: &Block, persisted: bool) -> Result<()> {
+        let mut serialized = vec![];
+        b.header.write(&mut serialized)?;
 
-    // fn fetch_block_tx_by_hash(&self, tx_id: &[u8]) -> Result<(ContractCall,
-    // u32, &[u8])> {     // Code for fetching block transaction by hash
-    // here     // ...
-    // }
+        self.inner
+            .put_cf(self.ledger_cf, b.header.hash, serialized)?;
 
-    fn fetch_block_hash_by_height(&self, height: u64) -> Result<&[u8]> {
-        Err(anyhow::Error::msg("message"))
+        anyhow::Ok(())
```

Comment:
Replace this `anyhow::Ok(())` with a simple `Ok(())`

Reply:
Done